storage/pebble/procedure/children.go

```diff
@@ -66,17 +65,13 @@ func IndexNewBlock(blockID flow.Identifier, parentID flow.Identifier) func(*badg
 		// adding the new block to be another child of the parent
 		childrenIDs = append(childrenIDs, blockID)
 
-		// saving the index
-		err = saveIndex(parentID, childrenIDs)(tx)
-		if err != nil {
-			return fmt.Errorf("could not update children index: %w", err)
-		}
-
-		return nil
+		// TODO: use transaction to avoid race condition
```

Comment:
⚠️ The method is not concurrent-safe at the database level. If two blocks are inserted concurrently, only one of them is stored. 

We are relying on the consensus algorithm to ensure the method is called in a single thread, which it does right now.

Reply:
regarding [your comment](https://github.com/onflow/flow-go/pull/6197/files#r1674455735) I think using a transaction provides inadequate guarantees for concurrent access. Please see [my comment](url). 

In my opinion, the only way to make this code safe is to delegate the sole authority for writing those indices to one single component. Consequently, this code here should be part of that higher-level component and not exported.

Reply:
## ⚠️ 

> We are relying on the consensus algorithm to ensure the method is called in a single thread, which it does right now.

I am not sure that is correct ... at least it is not apparent, so we have notable surface to accidentally break this in the future: 
* The compliance engine calls `State.Extend` https://github.com/onflow/flow-go/blob/51969ecf92ea9775622890d6be9be223f6f7a593/engine/consensus/compliance/core.go#L372
* The consensus builder is calling `State.Extend` https://github.com/onflow/flow-go/blob/51969ecf92ea9775622890d6be9be223f6f7a593/module/builder/consensus/builder.go#L147

Compliance engine and Block builder have different worker threads I think. Both call `Extend`, which in turn calls [`Mutator.insert`](https://github.com/onflow/flow-go/blob/51969ecf92ea9775622890d6be9be223f6f7a593/state/protocol/pebble/mutator.go#L217), which [calls `IndexNewBlock`](https://github.com/onflow/flow-go/blob/51969ecf92ea9775622890d6be9be223f6f7a593/state/protocol/pebble/mutator.go#L563). So we already have concurrent access in my opinion!